### ISSUE TITLE
organize imports action

### ISSUE THREAD

**Issue by GopherJ** [rust-analyzer/rust-analyzer#5131]
It'll be good to have this feature so that we can remove unused imports

**Comment by matklad** [comment]
cc https://github.com/rust-analyzer/rust-analyzer/issues/3301

**Comment by TonalidadeHidrica** [comment]
Actually, there are several operations I expect for "organize import" feature:

* Remove **all** redundant imports, and remove curly brackets if unneeded
* Merge **all** imports as much as possible, or split **all** imports, one identifier for one line
* Sort (and maybe insert appropriate blank lines between groups) the imports (the first of which can be already done by `cargo fmt`)

**Comment by branpk** [comment]
It would also be nice if local imports could be canonicalized, so that it consistently uses relative or absolute paths (or has a rule for determining which to use).

**Comment by dnut** [comment]
I would follow the approach intellij uses, with imports grouped by standard library, external, and local.

**Comment by aloucks** [comment]
I like the suggestions here (merging and sorting), but removal of unused imports is by far the what I'm missing the most.

**Comment by ocheret** [comment]
Is there any ongoing work on this issue?

**Comment by Ciel-MC** [comment]
There is already diagnostic info and quick fix for unused imports, what is stopping this from happening?

**Comment by Veykril** [comment]
A person willing to implement this (also not r-a itself does not yet diagnose unused imports, that's coming from the cargo checks)

**Comment by dnut** [comment]
Clippy can remove unused imports:
```
cargo clippy --fix
```

There should be a way to bind this to a keyboard shortcut. Some ideas: https://stackoverflow.com/questions/52786022/shortcut-for-running-terminal-command-in-vs-code

Caveats:
- applies to the entire workspace
- changes a bunch of things that rust-analyzer does not change
- does not work if you have local changes that have not been committed.

To deal with the last issue, my typical approach is to stage all my changes, so they are isolated from any additional changes. Then I run `cargo clippy --fix --allow-staged`. Then I review the unstaged changes from clippy. If it looks bad, I can revert them in isolation.

It would be ideal if rust-analyzer could clean up unused imports, but at least we have an alternative for now.

**Comment by Ciel-MC** [comment]
> Clippy can remove unused imports:
> 
> ```
> cargo clippy --fix
> ```
> 
> There should be a way to bind this to a keyboard shortcut. Some ideas: https://stackoverflow.com/questions/52786022/shortcut-for-running-terminal-command-in-vs-code
> 
> Caveats:
> 
>     * applies to the entire workspace
> 
>     * changes a bunch of things that rust-analyzer does not change
> 
>     * does not work if you have local changes that have not been committed.
> 
> 
> To deal with the last issue, my typical approach is to stage all my changes, so they are isolated from any additional changes. Then I run `cargo clippy --fix --allow-staged`. Then I review the unstaged changes from clippy. If it looks bad, I can revert them in isolation.
> 
> It would be ideal if rust-analyzer could clean up unused imports, but at least we have an alternative for now.

I am aware of cargo/clippy fix, but I don't want it to potentially nuke anything, but I couldn't figure out how to only fix a certain thing(unused imports), binding commands and stuff is the easiest part because... (neo)vim, lol.

**Comment by matklad** [comment]
Yeah, it feels like this should be doable. This is going to be a pretty large wad of code to add, but it seems to me that, at this point, we probably a reasonably precise analysis to make this worthwhile. 

Some general pointers: 

* this probably wants to be an assist, there's a doc comment about what it is [here](https://github.com/rust-lang/rust-analyzer/blob/62c81d62932c458e23975191ecc124916be0b35e/crates/ide-assists/src/lib.rs)
* [auto import](https://github.com/rust-lang/rust-analyzer/blob/62c81d62932c458e23975191ecc124916be0b35e/crates/ide-assists/src/handlers/auto_import.rs) might be a good read to learn about related machinery
* Similarly, [usage search](https://github.com/rust-lang/rust-analyzer/blob/master/crates/ide-db/src/search.rs) is probably something we'll use here. 

I _think_ this feature can work in one of two ways:

* In approach one, we record which imports do we use during analysis. So, eg, [`Body`](https://github.com/rust-lang/rust-analyzer/blob/62c81d62932c458e23975191ecc124916be0b35e/crates/hir-def/src/body.rs#L34) would get a field with the used imports, which is populated during lowering. import optimization than looks at all file's bodies (and whatever else can use imports), and computes the set of used things.
* Alternative approach would be to work from search. If you have `use foo::Bar`, and `Bar` is a struct, and the file _textually_ does not include the string `Bar`, then we can conclude that `Bar` is unused. If the `Bar` is a trait, then we should also look for mentions of trait's items.

Abstractly, I like the second approach much more, as it allows us to be _significantly_ more lazy. For all unused imports, we can probably say that they are unused following the parse, and for used stuff it's enough to have just one confirmed used to rule out a false negative. 

What makes me hesitant to try the second approach are macros. To _textually_ find all uses, we have to expand the macros... But this is something our search infra has to deal with anyway. 

So, yeah, my prior here is that building the feature on top of search is the way to go.

**Comment by matklad** [comment]
Ah, I realized that we already have a model of this workflow in the code, that's "remove unused param" assist: 

https://github.com/rust-lang/rust-analyzer/blob/62c81d62932c458e23975191ecc124916be0b35e/crates/ide-assists/src/handlers/remove_unused_param.rs

So I think import optimization should have roughly the same skeleton, with just a tiny bit more meat on it :D

**Comment by obsgolem** [comment]
I want to give this one a shot. If it turns out to be beyond me or I don't have the time I will give an update.

**Comment by obsgolem** [comment]
MR up!

**Comment by lem0nify** [comment]
when?

**Comment by igorakkerman** [comment]
I understand that the action is not fully completed yet. But is there a way to remove (all) unused imports at this point already (in VS Code) from any cursor position in the code and without scrolling?

The least disruptive solution I found so far is to open the problems view and to apply the assist on each problem.

**Comment by ospfranco** [comment]
![CleanShot 2024-06-04 at 08 54 10](https://github.com/rust-lang/rust-analyzer/assets/1634213/406f5993-b018-4293-a8fd-1633bd25b40e)

TIL you can half-way get this functionality. You just need to select the entire imports and then do `⌘` + `.` and select remove unused imports

**Comment by evrensen467** [comment]
any update on this? I'm so used to doing ``cmd`` + ``shift`` + ``o`` (which is the shortcut to the vscode "Organize Imports" command) to organize my imports in other languages like python, in rust it seems like this is already happening when you format the document but ideally I'd like to have an option to separate these actions.

**Comment by nmay231** [comment]
@nervenes Could you clarify what you mean by "this is already happening when you format the document". I'm pressing `ctrl+shift+I` and still have unused imports. Yes, selecting the imports and pressing `ctrl+.` brings up the option, but similar to @igorakkerman 's request I want to be able to do this without scrolling to the top each and every time.

Is there a config in vscode or `Cargo.toml` that you need to enable to get this to work? Many thanks.

**Comment by yuhr** [comment]
How can I make “Remove all the unused imports” happen when running “Format Document”, so that I can also run it on save? Is there any option to enable it on format?

**Comment by liamwh** [comment]
Bump, what is holding this up from being implemented? What's still needs to be done?

**Comment by ThiefMaster** [comment]
Yeah, what's the state of this? I wanted to try out the different values for the `rust-analyzer.imports.granularity.group` setting but I cannot find any way to actually apply it... what am I missing?!

**Comment by ShoyuVanilla** [comment]
> Yeah, what's the state of this? I wanted to try out the different values for the `rust-analyzer.imports.granularity.group` setting but I cannot find any way to actually apply it... what am I missing?!

[That option](https://github.com/rust-lang/rust-analyzer/issues/5131#issuecomment-3333294443) works for auto imports, like when you are auto-completing `Mutex` and rust-analyzer automatically inserting `use std::sync::Mutex`. I think that option would be applied to this feature but we don't have this yet. For state of this, well, I guess there's no major blockers for this but no one has worked on actually implementing this so far?

**Comment by chrs-b** [comment]
I'm using Zed (big recommendation) and it has a "Remove all unused imports" code action that works perfectly well.

**Comment by ForwardFeed** [comment]
As a workaround I use `cargo fix` which does remove all unused imports, maybe it does other things which you may not want, but that does the trick for me.

**Comment by nielsle** [comment]
Cargo-reorder claims to be able to reorder and group imports.  It looks very promising.

 https://github.com/CallOrRet/cargo-reorder
